galley/pkg/config/analysis/analyzers/virtualservice/matches.go

```diff
@@ -65,149 +63,25 @@ func (ma *MatchesAnalyzer) analyzeVirtualService(r *resource.Instance, c analysi
-			continue
-		}
-
-		duplicateMatches := 0
-		for matchn, match := range route.Match {
-			dupn, ok := matchesEncountered[asJSON(match)]
-			if ok {
-				c.Report(collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
-					msg.NewVirtualServiceIneffectiveMatch(r, routeName(route, rulen), requestName(match, matchn), routeName(routes[dupn], dupn)))
-				duplicateMatches++
-			} else {
-				matchesEncountered[asJSON(match)] = rulen
-			}
-		}
-		if duplicateMatches == len(route.Match) {
-			c.Report(collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
-				msg.NewVirtualServiceUnreachableRule(r, routeName(route, rulen), "all matches used by prior rules"))
-		}
-	}
-}
-
-// NOTE: This method identical to analyzeUnreachableHTTPRules.
-func (ma *MatchesAnalyzer) analyzeUnreachableTLSRules(r *resource.Instance, c analysis.Context, routes []*v1alpha3.TLSRoute) {
-	matchesEncountered := make(map[string]int)
-	emptyMatchEncountered := -1
-	for rulen, route := range routes {
-		if len(route.Match) == 0 {
-			if emptyMatchEncountered >= 0 {
-				c.Report(collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
-					msg.NewVirtualServiceUnreachableRule(r,
-						routeName(route, rulen), "only the last rule can have no matches"))
-			}
-			emptyMatchEncountered = rulen
-			continue
-		}
-
-		duplicateMatches := 0
-		for matchn, match := range route.Match {
-			dupn, ok := matchesEncountered[asJSON(match)]
-			if ok {
-				c.Report(collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
-					msg.NewVirtualServiceIneffectiveMatch(r, routeName(route, rulen), requestName(match, matchn), routeName(routes[dupn], dupn)))
-				duplicateMatches++
-			} else {
-				matchesEncountered[asJSON(match)] = rulen
-			}
-		}
-		if duplicateMatches == len(route.Match) {
-			c.Report(collections.IstioNetworkingV1Alpha3Virtualservices.Name(),
-				msg.NewVirtualServiceUnreachableRule(r, routeName(route, rulen), "all matches used by prior rules"))
-		}
-	}
-}
-
-func asJSON(data interface{}) string {
-	// Remove the name, so we can create a serialization that only includes traffic routing config
-	switch mr := data.(type) {
-	case *v1alpha3.HTTPMatchRequest:
-		if mr.Name != "" {
-			unnamed := *mr
-			unnamed.Name = ""
-			data = &unnamed
-		}
-	}
-
-	b, err := json.Marshal(data)
-	if err != nil {
-		return err.Error()
-	}
-	return string(b)
-}
-
-func routeName(route interface{}, routen int) string {
-	switch r := route.(type) {
-	case *v1alpha3.HTTPRoute:
-		if r.Name != "" {
-			return fmt.Sprintf("%q", r.Name)
-		}
-
-		// TCP and TLS routes have no names
-	}
-
-	return fmt.Sprintf("#%d", routen)
-}
-
-func requestName(match interface{}, matchn int) string {
-	switch mr := match.(type) {
-	case *v1alpha3.HTTPMatchRequest:
-		if mr.Name != "" {
-			return fmt.Sprintf("%q", mr.Name)
-		}
-
-		// TCP and TLS matches have no names
-	}
-
-	return fmt.Sprintf("#%d", matchn)
+	matches.AnalyzeUnreachableHTTPRules(vs.Http, func(ruleno, reason string) {
```

Comment:
why do we still need it here? We report all warnings from `pkg/config/validation/` in analyzer as well, so it seems redundant? I suspect you would actually get two warnings now?

Reply:
We do.  I left this dangling because I couldn't decide if it is OK if we "retire" the codes IST0130 and IST0131 and issue IST0133 for both of those issues.